crates/symbolicator-service/src/download/gcs.rs

```diff
@@ -36,22 +36,38 @@ impl GcsDownloader {
     ///
     /// If the cache contains a valid token, then this token is returned. Otherwise, a new token is
     /// requested from GCS and stored in the cache.
-    async fn get_token(&self, source_key: &Arc<GcsSourceKey>) -> CacheContents<GcsToken> {
+    async fn get_token(&self, source_key: Option<&Arc<GcsSourceKey>>) -> CacheContents<GcsToken> {
         metric!(counter("source.gcs.token.access") += 1);
 
-        let init = Box::pin(async {
-            metric!(counter("source.gcs.token.computation") += 1);
-            let token = gcs::request_new_token(&self.client, source_key).await;
-            token.map_err(CacheError::from)
-        });
-        let replace_if =
-            |entry: &CacheContents<GcsToken>| entry.as_ref().map_or(true, |t| t.is_expired());
-
-        self.token_cache
-            .entry_by_ref(source_key)
-            .or_insert_with_if(init, replace_if)
-            .await
-            .into_value()
+        match source_key {
+            Some(source_key) => {
+                let init = Box::pin(async {
+                    metric!(counter("source.gcs.token.computation") += 1);
+                    let token = gcs::request_new_token(&self.client, source_key).await;
+                    token.map_err(CacheError::from)
+                });
+                let replace_if = |entry: &CacheContents<GcsToken>| {
+                    entry.as_ref().map_or(true, |t| t.is_expired())
+                };
+
+                self.token_cache
+                    .entry_by_ref(source_key)
+                    .or_insert_with_if(init, replace_if)
+                    .await
+                    .into_value()
+            }
+            None => {
+                if let Ok(token_provider) = token_provider().await {
```

Comment:
Can't this provider be a field on the Downloader, then you might not need to cache the provider globally.